zarr/tests/test_codecs_v3.py

```diff
@@ -0,0 +1,989 @@
+from __future__ import annotations
+
+import json
+from typing import Iterator, List, Literal, Optional
+from attr import frozen
+
+import numpy as np
+import pytest
+import zarr
+from zarr.v3 import codecs
+from zarr.v3.array import Array, AsyncArray
+from zarr.v3.common import Selection
+from zarr.v3.indexing import morton_order_iter
+from zarr.v3.metadata import CodecMetadata, ShardingCodecIndexLocation, runtime_configuration
+
+from zarr.v3.store import MemoryStore, Store
+
+
+@frozen
+class _AsyncArrayProxy:
+    array: AsyncArray
+
+    def __getitem__(self, selection: Selection) -> _AsyncArraySelectionProxy:
+        return _AsyncArraySelectionProxy(self.array, selection)
+
+
+@frozen
+class _AsyncArraySelectionProxy:
+    array: AsyncArray
+    selection: Selection
+
+    async def get(self) -> np.ndarray:
+        return await self.array.getitem(self.selection)
+
+    async def set(self, value: np.ndarray):
+        return await self.array.setitem(self.selection, value)
```

Comment:
@jhamman @d-v-b What do you think about this making it into the AsyncArray? It allows for writing something like this
```
await _AsyncArrayProxy(array)[:10, 10:].get()
```
instead of 
```
await array.getitem((slice(None, 10), slice(10, None)))
```